### PR TITLE
[SwiftUI] Support presenting JavaScript dialogs

### PR DESCRIPTION
**Pull request by rr-codes**
#### e2eb78275e5ee653431c9dedd94138f7c6efe03b
<pre>
[SwiftUI] Support presenting JavaScript dialogs
<a href="https://bugs.webkit.org/show_bug.cgi?id=284402">https://bugs.webkit.org/show_bug.cgi?id=284402</a>
<a href="https://rdar.apple.com/141108805">rdar://141108805</a>

Reviewed by Wenson Hsieh.

Add the ability to support JavaScript alerts, confirmation prompts, text input prompts, and file input.

* Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift: Added.
(WKUIDelegateAdapter.webView(_:runJavaScriptAlertPanelWithMessage:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runJavaScriptConfirmPanelWithMessage:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runJavaScriptTextInputPanelWithPrompt:defaultText:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runOpenPanelWith:initiatedByFrame:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift: Added.
(DialogPresenting.handleJavaScriptAlert(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptConfirm(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(DialogPresenting.handleFileInputPrompt(_:initiatedBy:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Source/BrowserViewModel.swift:
(hash(into:)):
(owner):
(handleJavaScriptAlert(_:initiatedBy:)):
(handleJavaScriptConfirm(_:initiatedBy:)):
(handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(handleFileInputPrompt(_:initiatedBy:)):
(BrowserViewModel.currentDialog):
(BrowserViewModel.currentFilePicker):
(BrowserViewModel.didImportFiles(_:any:)):
* Tools/SwiftBrowser/Source/ContentView.swift:
(DialogActionsView.body):
(DialogMessageView.body):
(ContentView.body):

Canonical link: <a href="https://commits.webkit.org/287773@main">https://commits.webkit.org/287773@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/12ad3f84b251c72d9ec696f605985d02ee5c159e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80778 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/300 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34714 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/31762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8098 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/85305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/31762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83847 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/73530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/72 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/27690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/30219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/28222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/8005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/8182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/69366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12528 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7967 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/7806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/11325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9612 "Built successfully") | | |

### CODE REVIEW

File: Tools/SwiftBrowser/Source/ContentView.swift [hunk 1]
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
...
+        switch dialog.configuration {
+        case let .alert(_, dismissAlert):
+            Button("Dismiss") {
+                dismissAlert()
+            }
+
+        case let .confirm(_, dismissConfirm):
+            Button("OK") {
+                dismissConfirm(.ok)
+            }
```
Comment (lilyspiniolas) [inline Tools/SwiftBrowser/Source/ContentView.swift:79]: Should `.keyboardShortcut(.defaultAction)` be applied here too?

File: Tools/SwiftBrowser/Source/ContentView.swift [hunk 2]
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
...
+            _promptText = State(initialValue: defaultText)
+        }
+    }
+
+    var body: some View {
+        switch dialog.configuration {
+        case let .alert(_, dismissAlert):
+            Button("Dismiss") {
+                dismissAlert()
+            }
```
Comment (lilyspiniolas) [inline Tools/SwiftBrowser/Source/ContentView.swift]: nit: The current button for JS alerts is "Close," parity might be good here? Also not sure if it matters

File: Tools/SwiftBrowser/Source/ContentView.swift (hunk 1 above)
Comment (rr-codes) [reply Tools/SwiftBrowser/Source/ContentView.swift:79]: it's not needed, since if there are only buttons involved, SwiftUI can figure out which one is the default if the other is cancel.

File: Tools/SwiftBrowser/Source/ContentView.swift (hunk 2 above)
Comment (rr-codes) [reply Tools/SwiftBrowser/Source/ContentView.swift]: ya it doesn't really matter since this is just a test app, but I can change it

File: Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+#if ENABLE_SWIFTUI && compiler(>=6.0)
+
+import Foundation
+internal import WebKit_Private
+
+fileprivate struct DefaultDialogPresenting: DialogPresenting {
```
Comment (axiixc) [inline Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift]: Nit `fileprivate` is just `private` at top level scope usually.

File: Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+@_spi(Private)
+public extension DialogPresenting {
+    @MainActor
+    func handleJavaScriptAlert(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async {
+    }
+
+    @MainActor
+    func handleJavaScriptConfirm(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async -> WebPage_v0.JavaScriptConfirmResult {
+        .cancel
```
Comment (axiixc) [inline Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:73]: I'd probably prefer to write out `return` for a full method like this. But that's really just a stylistic thing to decide on.

File: Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    // MARK: Dialog presentation
+
+    func webView(_ webView: WKWebView, runJavaScriptAlertPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo) async {
+        await dialogPresenter.handleJavaScriptAlert(message: message, initiatedBy: .init(frame))
+    }
+
+    func webView(_ webView: WKWebView, runJavaScriptConfirmPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo) async -> Bool {
+        let result = await dialogPresenter.handleJavaScriptConfirm(message: message, initiatedBy: .init(frame))
+        return switch result {
+        case .ok:
```
Comment (axiixc) [inline Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift]: Since you are using implicit returns already, you could also shorten this to one line. That's kinda been my logic for these:
- Multiple lines, more complex, write `return`
- Single line, simpler expression, elide it

¯\\\_(ツ)\_/¯

File: Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+#if ENABLE_SWIFTUI && compiler(>=6.0)
+
+import Foundation
+
+// MARK: Supporting types
+
+extension WebPage_v0 {
+    public enum JavaScriptConfirmResult: Sendable {
```
Comment (axiixc) [inline Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift]: IIRC `enum`s are also implicitly `Hashable`, even as `public`, so you might want to write that down to be explicit. This is a bit of an outlier for Swift's general philosophy on API, and can bite you if were to add a new non-Hashable case like `case foo(OopsType)`. If you write down `: Hashable` now then you'll at least turn that into a build time error in the future (though we should also be running the ABI checker which would catch this _eventually_ too).

File: Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    @MainActor
+    func handleJavaScriptPrompt(message: String, defaultText: String?, initiatedBy frame: WebPage_v0.FrameInfo) async -> WebPage_v0.JavaScriptPromptResult
+
+    @MainActor
+    func handleFileInputPrompt(parameters: WKOpenPanelParameters, initiatedBy frame: WebPage_v0.FrameInfo) async -> WebPage_v0.FileInputPromptResult
+}
+
+// MARK: Default implementation
+
+@_spi(Private)
```
Comment (axiixc) [inline Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift]: I think this is technically unnecessary (the type itself is already SPI so shouldn't all extensions on it?)

File: Source/WebKit/UIProcess/API/Swift/WebPage.swift [hunk 8]
```diff
@@ -34,23 +34,33 @@ public class WebPage_v0 {
         WKWebView.handlesURLScheme(scheme)
     }
 
-    private init(configuration: Configuration, _navigationDecider navigationDecider: (any NavigationDeciding)?) {
-        self.configuration = configuration
+    private init(_configuration: Configuration, _navigationDecider navigationDecider: (any NavigationDeciding)?, _dialogPresenter dialogPresenter: (any DialogPresenting)?) {
```
Comment (axiixc) [inline Source/WebKit/UIProcess/API/Swift/WebPage.swift:37]: Curious about the `_`s in an already `private` method? Is this to avoid ambiguous overloads?

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift [hunk 9]
```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
...
+        let allowsMultipleSelection: Bool
+        let allowsDirectories: Bool
+        let completion: (WebPage_v0.FileInputPromptResult) -> Void
+    }
+
+    weak var owner: BrowserViewModel? = nil
+
+    func handleJavaScriptAlert(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async {
+        await withCheckedContinuation { continuation in
+            owner?.currentDialog = Dialog(configuration: .alert(message) {
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/BrowserViewModel.swift]: Super nit, but you should be able to just pass this as an unapplied function rather than making a closure just to call it like a thunk:
```suggestion
            owner?.currentDialog = Dialog(configuration: .alert(message, continuation.resume))
```

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift [hunk 10]
```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
...
+                continuation.resume()
+            })
+        }
+    }
+
+    func handleJavaScriptConfirm(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async -> WebPage_v0.JavaScriptConfirmResult {
+        await withCheckedContinuation { continuation in
+            owner?.currentDialog = Dialog(configuration: .confirm(message) {
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/BrowserViewModel.swift]: And likewise
```suggestion
            owner?.currentDialog = Dialog(configuration: .confirm(message, continuation.resume(returning:)))
```

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift [hunk 11]
```diff
@@ -67,6 +137,38 @@ final class BrowserViewModel {
...
+
+    var isPresentingDialog = false {
+        didSet {
+            if !isPresentingDialog {
+                currentDialog = nil
+            }
+        }
+    }
+
+    @ObservationIgnored
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/BrowserViewModel.swift]: Why ignored?

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift [hunk 12]
```diff
@@ -106,4 +208,18 @@ final class BrowserViewModel {
...
     }
+
+    func didImportFiles(result: Result<[URL], any Error>) {
+        precondition(currentFilePicker != nil)
+
+        switch result {
+        case let .success(urls):
+            currentFilePicker!.completion(.ok(urls))
+
+        case .failure(_):
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/BrowserViewModel.swift]: You can also drop the `()` entirely then:
```suggestion
        case .failure:
```

File: Tools/SwiftBrowser/Source/ContentView.swift [hunk 13]
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
...
+
+        if case let .prompt(_, defaultText, _) = dialog.configuration, let defaultText {
+            _promptText = State(initialValue: defaultText)
+        }
+    }
+
+    var body: some View {
+        switch dialog.configuration {
+        case let .alert(_, dismissAlert):
+            Button("Dismiss") {
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/ContentView.swift]:
```suggestion
            Button("Dismiss", action: dismissAlert)
```

File: Tools/SwiftBrowser/Source/ContentView.swift (hunk 1 above)
Comment (axiixc) [reply Tools/SwiftBrowser/Source/ContentView.swift:79]: Yeah, if you had more than one non-cancel non-destructive role button you'd probably need that to tie break otherwise.

File: Tools/SwiftBrowser/Source/ContentView.swift [hunk 14]
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
...
+        }
+    }
+}
+
+fileprivate struct DialogMessageView: View {
+    let dialog: DialogPresenter.Dialog
+
+    var body: some View {
+        switch dialog.configuration {
+        case let .alert(message, _):
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/ContentView.swift:105]: This is just sample code, so N.B.D., but another way to handle this would be making a `var message: String { … }` extension on the enum so that you could write `Text(dialog.configuration.message)` here. That's mostly isomorphic, but it does simplify your view tree. That only applies because the types here are all otherwise identical, so if there were any other customizations I probably wouldn't recommend it. Just throwing it out.

**Review by axiixc — APPROVED** [review]

File: Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift (hunk 5 above)
Comment (rr-codes) [reply Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift]: so it would be like

```
switch await dialogPresenter... { 
```

?

File: Source/WebKit/UIProcess/API/Swift/WebPage.swift (hunk 8 above)
Comment (rr-codes) [reply Source/WebKit/UIProcess/API/Swift/WebPage.swift:37]: ya, the initializers beneath it caused this to become ambiguous

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift (hunk 11 above)
Comment (rr-codes) [reply Tools/SwiftBrowser/Source/BrowserViewModel.swift]: I figured that it wasn't necessary since the "isPresentingDialog" property takes care of the observable-ness needed

File: Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift (hunk 5 above)
Comment (axiixc) [reply Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift]: Sorry, I was implying:
```swift
let result = await dialogPresenter.handleJavaScriptConfirm(message: message, initiatedBy: .init(frame))

return switch result {
case .ok: true
case .cancel: false
}
```

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift [hunk 15]
```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
...
+            case alert(String, @Sendable () -> Void)
+            case confirm(String, @Sendable (WebPage_v0.JavaScriptConfirmResult) -> Void)
+            case prompt(String, defaultText: String?, @Sendable (WebPage_v0.JavaScriptPromptResult) -> Void)
+        }
+
+        let id = UUID()
+        let configuration: Configuration
+
+        func hash(into hasher: inout Hasher) {
+            self.id.hash(into: &hasher)
```
Comment (axiixc) [inline Tools/SwiftBrowser/Source/BrowserViewModel.swift]: Almost missed this, but what you actually want is
```suggestion
           hasher.combine(id)
```

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift (hunk 10 above)
Comment (rr-codes) [reply Tools/SwiftBrowser/Source/BrowserViewModel.swift]: When I try this, I get

```
Converting a value of type '@Sendable (sending WebPage_v0.JavaScriptConfirmResult) -> ()' to type '@Sendable (WebPage_v0.JavaScriptConfirmResult) -> Void' risks causing data races
```

Comment (rr-codes) [reply Tools/SwiftBrowser/Source/BrowserViewModel.swift]: I made the enum associated value have a `sending` parameter and that fixed it

**Review by whsieh — APPROVED** [review]

File: Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift (hunk 4 above)
Comment (rr-codes) [reply Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:73]: I think I use implicit returns in other cases, so I think I'll leave it as-is for now

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift (hunk 10 above)
Comment (axiixc) [reply Tools/SwiftBrowser/Source/BrowserViewModel.swift]: Hmm, I was expecting this to just convert, but some of the concurrency annotations are more strict (ex: you can generally pass a `(Int?) -> Void` to an `(Int) -> Void` because it's a subtype).

File: Tools/SwiftBrowser/Source/BrowserViewModel.swift (hunk 11 above)
Comment (axiixc) [reply Tools/SwiftBrowser/Source/BrowserViewModel.swift]: I'd personally avoid overeager optimizations. If you capture two dependencies and they both change in one render cycle that's still one re-render. Conversely if you only capture one and the other changes you'll have a correctness bug.

`@ObservationIgnored` is good when you want to semantically have no side effects on a read because it's transparent to the caller for whatever reason. Usually I use this with memoized subscripts or methods, because the fact that you've cached a new value doesn't mean you want a dependency any time you cache any other new value, unless of course you have a debug view showing the contents of the cache in which case it _would_ be wrong.
